rbd.c

```diff
@@ -49,9 +64,254 @@ struct rbd_aio_cb {
+	rados_shutdown(state->cluster);
+
+	state->cluster = NULL;
+	state->io_ctx = NULL;
+	state->image = NULL;
+
+	pthread_spin_lock(&state->lock);
+	state->state = TCMU_RBD_CLOSED;
+	pthread_spin_unlock(&state->lock);
+}
+
+static int tcmu_rbd_image_open(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	int ret;
+
+	pthread_spin_lock(&state->lock);
+	if (state->state == TCMU_RBD_OPENED) {
+		tcmu_dev_dbg(dev, "skipping open. Already opened\n");
+		pthread_spin_unlock(&state->lock);
+		return 0;
+	}
+
+	if (state->state != TCMU_RBD_CLOSED) {
+		tcmu_dev_dbg(dev, "skipping open. state %d\n", state->state);
+		pthread_spin_unlock(&state->lock);
+		return -EBUSY;
+	}
+	state->state = TCMU_RBD_OPENING;
+	pthread_spin_unlock(&state->lock);
+
+	/* TODO locking */
+	ret = rados_create(&state->cluster, NULL);
+	if (ret < 0) {
+		tcmu_dev_dbg(dev, "Could not create cluster. (Err %d)\n", ret);
+		goto set_closed;
+	}
+
+	/* Fow now, we will only read /etc/ceph/ceph.conf */
+	rados_conf_read_file(state->cluster, NULL);
+	rados_conf_set(state->cluster, "rbd_cache", "false");
+
+	ret = rados_connect(state->cluster);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not connect to cluster. (Err %d)\n",
+			     ret);
+		goto rados_shutdown;
+	}
+
+	ret = rados_ioctx_create(state->cluster, state->pool_name,
+				 &state->io_ctx);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not create ioctx for pool %s. (Err %d)\n",
+			     state->pool_name, ret);
+		goto rados_destroy;
+	}
+
+	ret = rbd_open(state->io_ctx, state->image_name, &state->image, NULL);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not open image %s. (Err %d)\n",
+			     state->image_name, ret);
+		goto rados_destroy;
+	}
+
+	pthread_spin_lock(&state->lock);
+	state->state = TCMU_RBD_OPENED;
+	pthread_spin_unlock(&state->lock);
+	return 0;
+
+rados_destroy:
+	rados_ioctx_destroy(state->io_ctx);
+	state->io_ctx = NULL;
+rados_shutdown:
+	rados_shutdown(state->cluster);
+	state->cluster = NULL;
+set_closed:
+	pthread_spin_lock(&state->lock);
+	state->state = TCMU_RBD_CLOSED;
+	pthread_spin_unlock(&state->lock);
+	return ret;
+}
+
+static int tcmu_rbd_image_reopen(struct tcmu_device *dev)
+{
+	int ret;
+
+	tcmu_dev_dbg(dev, "image reopen\n");
+	tcmu_rbd_image_close(dev);
+	ret = tcmu_rbd_image_open(dev);
+	tcmu_dev_dbg(dev, "image reopen got %d\n", ret);
+	return ret;
+}
+
+static int tcmu_rbd_lock(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	int ret = 0, attempts = 0;
+
+	/*
+	 * TODO: Add retry/timeout settings to handle windows/ESX.
+	 * Or, set to transitioning and grab the lock in the background.
+	 */
+	while (attempts++ < 5) {
+		ret = is_exclusive_lock_owner(dev);
+		if (ret == 1) {
+			ret = 0;
+			break;
+		} else if (ret == -ESHUTDOWN) {
+			ret = tcmu_rbd_image_reopen(dev);
+			continue;
+		} else if (ret < 0) {
+			sleep(1);
+			continue;
+		}
+
+		ret = rbd_lock_acquire(state->image, RBD_LOCK_MODE_EXCLUSIVE);
```

Comment:
For the last part, if I understand you it is handled. Is the concern we could end up with node1 and node2 getting a failover command, and they both end up returning success? If so this is handled. tcmu-runner will serialize failover commands sent to the same device on the same node.

If we had:
1. initiator sends failover command to node1 then node2, but node1 grabbed lock then node2.

The initiator side, would have marked the path to node1 as bad before sending a failover command to node2. The initiator side will only see the failover command response from node2.  The response from node1 will be dropped. When the initiator gets the success response from node2, it will send IO. When node1's path comes back the initiator side will resync its state and will see node1 does not have the lock, so it will not send IO to it.

2.  initiator sends failover command to node1 then node2, but node2 grabs the lock then node1.

In this case, the initiator side will mark node1 as bad like in case 1. It would then get a success response from node2 for its failover command. But, new IO commands sent from the initiator will fail because node2 had been blacklisted when node1 took the lock. The target side fails those commands with a special error code, and to the initiator it will similar to an implicit failover. The initiator would respond by sending another failover command to node2 and that will succeed and we can do IO again.

The current issue with this case is what I mentioned in my status the other day. The special error code gets dropped by the Lio kernel code and replaced with a error code some initiators treat as a hard error. Instead of sending another failover command, they offline the device.